delocate/tools.py

```diff
@@ -163,9 +176,197 @@ def parse_install_name(line):
+    return libname, compat_version, current_version
+
+
+_OTOOL_ARCHITECTURE_RE = re.compile(
+    r"^(?P<name>.*?)(?: \(architecture (?P<architecture>\w+)\))?:$"
+)
+"""Matches the library separater line in 'otool -L'-like outputs.
+
+Examples
+--------
+>>> match = _OTOOL_ARCHITECTURE_RE.match("example.so (architecture x86_64):")
+>>> match["name"]
+'example.so'
+>>> match["architecture"]
+'x86_64'
+>>> match = _OTOOL_ARCHITECTURE_RE.match("example.so:")
+>>> match["name"]
+'example.so'
+>>> match["architecture"] is None
+True
+"""
+
+
+def _parse_otool_listing(stdout: str) -> Dict[str, List[str]]:
+    '''Parse the output of otool lists.
+
+    Parameters
+    ----------
+    stdout : str
+        A decoded stdout of commands like 'otool -L' or 'otool -D' to be parsed.
+
+    Returns
+    -------
+    otool_out : dict of list of str
+        The dictionary key is the architecture, e.g. 'x86_64'.
+        If no architecture is parsed then the key is ''.
+        The values are a list of strings associated with the key.
+
+    Examples
+    --------
+    >>> _parse_otool_listing("""
+    ... example.so (architecture x86_64):
+    ... \titem_1
+    ... \titem_2
+    ... example.so (architecture arm64):
+    ... \titem_3
+    ... \titem_4
+    ... """)
+    {'x86_64': ['item_1', 'item_2'], 'arm64': ['item_3', 'item_4']}
+    >>> _parse_otool_listing("""
+    ... example.so:
+    ... \titem_1
+    ... """)
+    {'': ['item_1']}
+    >>> _parse_otool_listing("""
+    ... example.so (architecture x86_64):
+    ... example.so (architecture arm64):
+    ... """)
+    {'x86_64': [], 'arm64': []}
+    >>> _parse_otool_listing("")
+    Traceback (most recent call last):
+        ...
+    RuntimeError: Missing file/architecture header:...
+    >>> _parse_otool_listing("""
+    ... example.so (architecture arm64):
+    ... example.so (architecture arm64):
+    ... """)
+    Traceback (most recent call last):
+        ...
+    RuntimeError: Input has duplicate architectures for ...
+    '''
+    stdout = stdout.strip()
+    out: Dict[str, List[str]] = {}
+    lines = stdout.split("\n")
+    while lines:
+        # Detect and parse the name/arch header line.
+        match_arch = _OTOOL_ARCHITECTURE_RE.match(lines.pop(0))
+        if not match_arch:
+            raise RuntimeError(f"Missing file/architecture header:\n{stdout}")
+        current_arch: Optional[str] = match_arch["architecture"]
+        if current_arch is None:
+            current_arch = ""
+        if current_arch in out:
+            raise RuntimeError(
+                "Input has duplicate architectures for"
+                f" {current_arch!r}:\n{stdout}"
+            )
+        out[current_arch] = []
+        # Collect lines until the next header or the end.
+        while lines and not lines[0].endswith(":"):
+            out[current_arch].append(lines.pop(0).strip())
+    return out
+
+
+def _ignore_architectures(input: Dict[str, T]) -> T:
```

Comment:
Maybe `_check_ignore_archs`?   I missed, first time round, that this function raises for not-same values for different archs.

Reply:
I agree that the current name does not imply that some inputs will raise errors.  I'll change it.